README.rst
========================
Team and repository tags
========================

.. image:: http://governance.openstack.org/badges/deb-nova.svg
    :target: http://governance.openstack.org/reference/tags/index.html

.. Change things from this point on

OpenStack Nova
==============

OpenStack Nova provides a cloud computing fabric controller,
supporting a wide variety of compute technologies, including:
libvirt (KVM, Xen, LXC and more), Hyper-V, VMware, XenServer
and OpenStack Ironic.

OpenStack Nova is distributed under the terms of the Apache
License, Version 2.0. The full terms and conditions of this
license are detailed in the LICENSE file.

API
---

To learn how to use Nova's API, consult the documentation
available online at:

    http://developer.openstack.org/api-guide/compute/
    http://developer.openstack.org/api-ref/compute/

For more information on OpenStack APIs, SDKs and CLIs,
please see:

    http://www.openstack.org/appdev/
    http://developer.openstack.org/

Operators
---------

To learn how to deploy and configure OpenStack Nova, consult the
documentation available online at:

    http://docs.openstack.org

For information about the different compute (hypervisor) drivers
supported by Nova, please read:

   http://docs.openstack.org/developer/nova/feature_classification.html

In the unfortunate event that bugs are discovered, they should
be reported to the appropriate bug tracker. If you obtained
the software from a 3rd party operating system vendor, it is
often wise to use their own bug tracker for reporting problems.
In all other cases use the master OpenStack bug tracker,
available at:

   http://bugs.launchpad.net/nova

Developers
----------

For information on how to contribute to Nova, please see the
contents of the CONTRIBUTING.rst.

Any new code must follow the development guidelines detailed
in the HACKING.rst file, and pass all unit tests.

Further developer focused documentation is available at:

   http://docs.openstack.org/developer/nova/


debian/changelog
nova (2:14.0.0-4) unstable; urgency=medium

  [ David Rabel ]
  * Team upload.
  * Bump build dependency on openstack-pkg-tools (Closes: #858708, #858710).

  [ Thomas Goirand ]
  * CVE-2017-7214: apply upstream patch (Closes: 858568).

 -- Thomas Goirand <zigo@debian.org>  Sun, 02 Apr 2017 12:52:50 +0200

nova (2:14.0.0-3) unstable; urgency=medium

  [ Ondřej Nový ]
  * Bumped debhelper compat version to 10

  [ Thomas Goirand ]
  * Allow using SQLAlchemy >= 1.1.
  * Fix bug that prevented generating nova-spicehtml5proxy and nova-xenvncproxy
    .service files.

 -- Thomas Goirand <zigo@debian.org>  Fri, 09 Dec 2016 17:40:19 +0100

nova (2:14.0.0-2) unstable; urgency=medium

  * Patches migration script to clean-up build_requests on upgrades from
    Mitaka.

 -- Thomas Goirand <zigo@debian.org>  Wed, 19 Oct 2016 14:28:43 +0200

nova (2:14.0.0-1) unstable; urgency=medium

  * New upstream release.

 -- Thomas Goirand <zigo@debian.org>  Thu, 06 Oct 2016 17:22:58 +0200

nova (2:14.0.0~rc1-2) unstable; urgency=medium

  [ Ondřej Nový ]
  * d/s/options: extend-diff-ignore of .gitreview
  * d/control: Use correct branch in Vcs-* fields

  [ Thomas Goirand ]
  * Uploading to unstable.
  * Fixed oslotest EPOCH.

 -- Thomas Goirand <zigo@debian.org>  Wed, 28 Sep 2016 10:05:08 +0200

nova (2:14.0.0~rc1-1) experimental; urgency=medium

  * New upstream release.
  * Fixed (build-)depends for this release.
  * Using OpenStack's Gerrit as VCS URLs.
  * Points .gitreview to OpenStack's Gerrit packaging-deb.
  * Fixed nova.conf generation.
  * Blacklist failing test:
    - virt.libvirt.test_driver.LibvirtConnTestCase.test_spawn_with_config_drive
  * Added a nova-placement-api package & daemon.
  * Package /usr/bin/nova-policy in nova-common.
  * Added --parallel when running unit tests.
  * Revised debian/copyright holder list.

 -- Thomas Goirand <zigo@debian.org>  Wed, 14 Sep 2016 09:40:05 +0200

nova (2:14.0.0~b2-1) experimental; urgency=medium

  * New upstream release.
  * Fixed (build-)depends for this release.
  * Updated Danish translation of the debconf (Closes: #830630).

 -- Thomas Goirand <zigo@debian.org>  Mon, 11 Jul 2016 14:18:44 +0200

nova (2:14.0.0~b1-2) experimental; urgency=medium

  * Fixed correct path for $pybasedir.
  * Fixed bindir to /usr/bin.
  * Fixed use_neutron to True by default.
  * Fixed api_servers to http://localhost:9292 by default.
  * Rebuilt with openstack-pkg-tools >= 52~ to ensure ALTER TABLE is done
    correctly with backquotes for the db name.

 -- Thomas Goirand <zigo@debian.org>  Thu, 30 Jun 2016 08:29:14 +0000

nova (2:14.0.0~b1-1) experimental; urgency=medium

  * New upstream release.
  * Fixed (build-)depends for this release.
  * Update the namespace list when generating nova.conf.
  * Adds fix-requirements.txt.patch.

 -- Thomas Goirand <zigo@debian.org>  Wed, 08 Jun 2016 14:46:53 +0200

nova (2:13.0.0-3) unstable; urgency=medium

  [ Thomas Goirand ]
  * Updated Japanese debconf templates translation update (Closes: #820768).
  * Updated Dutch debconf templates translation (Closes: #822887).
  * Updated Brazilian Portuguese debconf templates (Closes: #824336).

  [ Ondřej Nový ]
  * Added Documentation to systemd units

 -- Thomas Goirand <zigo@debian.org>  Sat, 23 Apr 2016 18:56:19 +0200

nova (2:13.0.0-2) unstable; urgency=medium

  [ Ondřej Nový ]
  * Use /bin/sh as su shell in postinst script explicitly
  * Standards-Version is 3.9.8 now (no change)
  * Use /bin/sh as default shell for "nova" user

 -- Thomas Goirand <zigo@debian.org>  Thu, 14 Apr 2016 11:04:59 +0000

nova (2:13.0.0-1) unstable; urgency=medium

  * New upstream release.

 -- Thomas Goirand <zigo@debian.org>  Thu, 07 Apr 2016 18:17:11 +0200

nova (2:13.0.0~rc3-1) unstable; urgency=medium

  * New upstream release.
  * Uploading to unstable.
  * Updated ja.po debconf translations (Closes: #815955).

 -- Thomas Goirand <zigo@debian.org>  Tue, 05 Apr 2016 09:43:40 +0200

nova (2:13.0.0~rc1-4) experimental; urgency=medium

  * Standards-Version: 3.9.7 (no change)
  * Do not use keystone admin auth token anymore (and therefore, build-depends
    on openstack-pkg-tools >= 40~).

 -- Thomas Goirand <zigo@debian.org>  Tue, 29 Mar 2016 09:45:11 +0000

nova (2:13.0.0~rc1-3) experimental; urgency=medium

  * Fixed netron -> neutron in nova-common.postinst.in.

 -- Thomas Goirand <zigo@debian.org>  Tue, 29 Mar 2016 09:35:04 +0000

nova (2:13.0.0~rc1-2) experimental; urgency=medium

  * Added dbconfig-common management of the novaapi db.

 -- Thomas Goirand <zigo@debian.org>  Tue, 29 Mar 2016 08:46:01 +0000

nova (2:13.0.0~rc1-1) experimental; urgency=medium

  [ Ivan Udovichenko ]
  * d/patches: Add Install-missed-files.patch file.
  * d/patches/series: Remove outdated entry and add the new one.

  [ Thomas Goirand ]
  * New upstream release.
  * Fix [neutron]/auth_plugin -> [neutron]/auth_type fixup, to follow the
    changes in upstream generated config file.
  * Fixed (build-)depends for this release.

 -- Thomas Goirand <zigo@debian.org>  Mon, 21 Mar 2016 14:32:43 +0100

nova (2:13.0.0~b3-1) experimental; urgency=medium

  [ Ondřej Nový ]
  * Fixed VCS URLs (https).

  [ Thomas Goirand ]
  * New upstream release.
  * Fixed (build-)depends for this release.
  * Fix namespace list for generating nova.conf.

 -- Thomas Goirand <zigo@debian.org>  Fri, 04 Mar 2016 03:05:59 +0800

nova (2:13.0.0~b2-2) experimental; urgency=medium

  [ Ivan Udovichenko ]
  * d/rules: Binary for nova-objectstore service is not generated anymore.

  [ Thomas Goirand ]
  * Do not use -S flag of dpkg-parsechangelog: incompatible with Trusty.
  * Add build-depends-indep in Git to avoid crash during sphinx doc build.
  * Bump EPOCH to align with Ubuntu.
  * Stop using hand-crafted config file, and use the generated one.
  * Removed ec2 from possible API activation (support for that doesn't exist
    anymore).

 -- Thomas Goirand <zigo@debian.org>  Fri, 29 Jan 2016 06:08:52 +0000


nova/virt/fake.py
# Copyright 2010 United States Government as represented by the
# Administrator of the National Aeronautics and Space Administration.
# All Rights Reserved.
# Copyright (c) 2010 Citrix Systems, Inc.
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

"""
A fake (in-memory) hypervisor+api.

Allows nova testing w/o a hypervisor.  This module also documents the
semantics of real hypervisor connections.

"""

import contextlib

from oslo.config import cfg

from nova.compute import power_state
from nova.compute import task_states
from nova.console import type as ctype
from nova import db
from nova import exception
from nova.i18n import _
from nova.openstack.common import jsonutils
from nova.openstack.common import log as logging
from nova import utils
from nova.virt import diagnostics
from nova.virt import driver
from nova.virt import virtapi

CONF = cfg.CONF
CONF.import_opt('host', 'nova.netconf')

LOG = logging.getLogger(__name__)


_FAKE_NODES = None


def set_nodes(nodes):
    """Sets FakeDriver's node.list.

    It has effect on the following methods:
        get_available_nodes()
        get_available_resource
        get_host_stats()

    To restore the change, call restore_nodes()
    """
    global _FAKE_NODES
    _FAKE_NODES = nodes


def restore_nodes():
    """Resets FakeDriver's node list modified by set_nodes().

    Usually called from tearDown().
    """
    global _FAKE_NODES
    _FAKE_NODES = [CONF.host]


class FakeInstance(object):

    def __init__(self, name, state, uuid):
        self.name = name
        self.state = state
        self.uuid = uuid

    def __getitem__(self, key):
        return getattr(self, key)


class FakeDriver(driver.ComputeDriver):
    capabilities = {
        "has_imagecache": True,
        "supports_recreate": True,
        }

    """Fake hypervisor driver."""

    def __init__(self, virtapi, read_only=False):
        super(FakeDriver, self).__init__(virtapi)
        self.instances = {}
        self.host_status_base = {
          'vcpus': 100000,
          'memory_mb': 8000000000,
          'local_gb': 600000000000,
          'vcpus_used': 0,
          'memory_mb_used': 0,
          'local_gb_used': 100000000000,
          'hypervisor_type': 'fake',
          'hypervisor_version': utils.convert_version_to_int('1.0'),
          'hypervisor_hostname': CONF.host,
          'cpu_info': {},
          'disk_available_least': 500000000000,
          'supported_instances': [(None, 'fake', None)],
          }
        self._mounts = {}
        self._interfaces = {}
        if not _FAKE_NODES:
            set_nodes([CONF.host])

    def init_host(self, host):
        return

    def list_instances(self):
        return self.instances.keys()

    def list_instance_uuids(self):
        return [self.instances[name].uuid for name in self.instances.keys()]

    def plug_vifs(self, instance, network_info):
        """Plug VIFs into networks."""
        pass

    def unplug_vifs(self, instance, network_info):
        """Unplug VIFs from networks."""
        pass

    def spawn(self, context, instance, image_meta, injected_files,
              admin_password, network_info=None, block_device_info=None):
        name = instance['name']
        state = power_state.RUNNING
        fake_instance = FakeInstance(name, state, instance['uuid'])
        self.instances[name] = fake_instance

    def snapshot(self, context, instance, name, update_task_state):
        if instance['name'] not in self.instances:
            raise exception.InstanceNotRunning(instance_id=instance['uuid'])
        update_task_state(task_state=task_states.IMAGE_UPLOADING)

    def reboot(self, context, instance, network_info, reboot_type,
               block_device_info=None, bad_volumes_callback=None):
        pass

    @staticmethod
    def get_host_ip_addr():
        return '192.168.0.1'

    def set_admin_password(self, instance, new_pass):
        pass

    def inject_file(self, instance, b64_path, b64_contents):
        pass

    def resume_state_on_host_boot(self, context, instance, network_info,
                                  block_device_info=None):
        pass

    def rescue(self, context, instance, network_info, image_meta,
               rescue_password):
        pass

    def unrescue(self, instance, network_info):
        pass

    def poll_rebooting_instances(self, timeout, instances):
        pass

    def migrate_disk_and_power_off(self, context, instance, dest,
                                   flavor, network_info,
                                   block_device_info=None,
                                   timeout=0, retry_interval=0):
        pass

    def finish_revert_migration(self, context, instance, network_info,
                                block_device_info=None, power_on=True):
        pass

    def post_live_migration_at_destination(self, context, instance,
                                           network_info,
                                           block_migration=False,
                                           block_device_info=None):
        pass

    def power_off(self, instance, shutdown_timeout=0, shutdown_attempts=0):
        pass

    def power_on(self, context, instance, network_info, block_device_info):
        pass

    def soft_delete(self, instance):
        pass

    def restore(self, instance):
        pass

    def pause(self, instance):
        pass

    def unpause(self, instance):
        pass

    def suspend(self, instance):
        pass

    def resume(self, context, instance, network_info, block_device_info=None):
        pass

    def destroy(self, context, instance, network_info, block_device_info=None,
                destroy_disks=True, migrate_data=None):
        key = instance['name']
        if key in self.instances:
            del self.instances[key]
        else:
            LOG.warning(_("Key '%(key)s' not in instances '%(inst)s'") %
                        {'key': key,
                         'inst': self.instances}, instance=instance)

    def cleanup(self, context, instance, network_info, block_device_info=None,
                destroy_disks=True, migrate_data=None, destroy_vifs=True):
        pass

    def attach_volume(self, context, connection_info, instance, mountpoint,
                      disk_bus=None, device_type=None, encryption=None):
        """Attach the disk to the instance at mountpoint using info."""
        instance_name = instance['name']
        if instance_name not in self._mounts:
            self._mounts[instance_name] = {}
        self._mounts[instance_name][mountpoint] = connection_info

    def detach_volume(self, connection_info, instance, mountpoint,
                      encryption=None):
        """Detach the disk attached to the instance."""
        try:
            del self._mounts[instance['name']][mountpoint]
        except KeyError:
            pass

    def swap_volume(self, old_connection_info, new_connection_info,
                    instance, mountpoint, resize_to):
        """Replace the disk attached to the instance."""
        instance_name = instance['name']
        if instance_name not in self._mounts:
            self._mounts[instance_name] = {}
        self._mounts[instance_name][mountpoint] = new_connection_info

    def attach_interface(self, instance, image_meta, vif):
        if vif['id'] in self._interfaces:
            raise exception.InterfaceAttachFailed(
                    instance_uuid=instance['uuid'])
        self._interfaces[vif['id']] = vif

    def detach_interface(self, instance, vif):
        try:
            del self._interfaces[vif['id']]
        except KeyError:
            raise exception.InterfaceDetachFailed(
                    instance_uuid=instance['uuid'])

    def get_info(self, instance):
        if instance['name'] not in self.instances:
            raise exception.InstanceNotFound(instance_id=instance['name'])
        i = self.instances[instance['name']]
        return {'state': i.state,
                'max_mem': 0,
                'mem': 0,
                'num_cpu': 2,
                'cpu_time': 0}

    def get_diagnostics(self, instance_name):
        return {'cpu0_time': 17300000000,
                'memory': 524288,
                'vda_errors': -1,
                'vda_read': 262144,
                'vda_read_req': 112,
                'vda_write': 5778432,
                'vda_write_req': 488,
                'vnet1_rx': 2070139,
                'vnet1_rx_drop': 0,
                'vnet1_rx_errors': 0,
                'vnet1_rx_packets': 26701,
                'vnet1_tx': 140208,
                'vnet1_tx_drop': 0,
                'vnet1_tx_errors': 0,
                'vnet1_tx_packets': 662,
        }

    def get_instance_diagnostics(self, instance_name):
        diags = diagnostics.Diagnostics(state='running', driver='fake',
                hypervisor_os='fake-os', uptime=46664, config_drive=True)
        diags.add_cpu(time=17300000000)
        diags.add_nic(mac_address='01:23:45:67:89:ab',
                      rx_packets=26701,
                      rx_octets=2070139,
                      tx_octets=140208,
                      tx_packets = 662)
        diags.add_disk(id='fake-disk-id',
                       read_bytes=262144,
                       read_requests=112,
                       write_bytes=5778432,
                       write_requests=488)
        diags.memory_details.maximum = 524288
        return diags

    def get_all_bw_counters(self, instances):
        """Return bandwidth usage counters for each interface on each
           running VM.
        """
        bw = []
        return bw

    def get_all_volume_usage(self, context, compute_host_bdms):
        """Return usage info for volumes attached to vms on
           a given host.
        """
        volusage = []
        return volusage

    def get_host_cpu_stats(self):
        stats = {'kernel': 5664160000000L,
                'idle': 1592705190000000L,
                'user': 26728850000000L,
                'iowait': 6121490000000L}
        stats['frequency'] = 800
        return stats

    def block_stats(self, instance_name, disk_id):
        return [0L, 0L, 0L, 0L, None]

    def interface_stats(self, instance_name, iface_id):
        return [0L, 0L, 0L, 0L, 0L, 0L, 0L, 0L]

    def get_console_output(self, context, instance):
        return 'FAKE CONSOLE OUTPUT\nANOTHER\nLAST LINE'

    def get_vnc_console(self, context, instance):
        return ctype.ConsoleVNC(internal_access_path='FAKE',
                                host='fakevncconsole.com',
                                port=6969)

    def get_spice_console(self, context, instance):
        return ctype.ConsoleSpice(internal_access_path='FAKE',
                                  host='fakespiceconsole.com',
                                  port=6969,
                                  tlsPort=6970)

    def get_rdp_console(self, context, instance):
        return ctype.ConsoleRDP(internal_access_path='FAKE',
                                host='fakerdpconsole.com',
                                port=6969)

    def get_serial_console(self, context, instance):
        return ctype.ConsoleSerial(internal_access_path='FAKE',
                                   host='fakerdpconsole.com',
                                   port=6969)

    def get_console_pool_info(self, console_type):
        return {'address': '127.0.0.1',
                'username': 'fakeuser',
                'password': 'fakepassword'}

    def refresh_security_group_rules(self, security_group_id):
        return True

    def refresh_security_group_members(self, security_group_id):
        return True

    def refresh_instance_security_rules(self, instance):
        return True

    def refresh_provider_fw_rules(self):
        pass

    def get_available_resource(self, nodename):
        """Updates compute manager resource info on ComputeNode table.

           Since we don't have a real hypervisor, pretend we have lots of
           disk and ram.
        """
        if nodename not in _FAKE_NODES:
            return {}

        dic = {'vcpus': 1,
               'memory_mb': 8192,
               'local_gb': 1028,
               'vcpus_used': 0,
               'memory_mb_used': 0,
               'local_gb_used': 0,
               'hypervisor_type': 'fake',
               'hypervisor_version': utils.convert_version_to_int('1.0'),
               'hypervisor_hostname': nodename,
               'disk_available_least': 0,
               'cpu_info': '?',
               'supported_instances': jsonutils.dumps([(None, 'fake', None)])
              }
        return dic

    def ensure_filtering_rules_for_instance(self, instance_ref, network_info):
        return

    def get_instance_disk_info(self, instance_name, block_device_info=None):
        return

    def live_migration(self, context, instance_ref, dest,
                       post_method, recover_method, block_migration=False,
                       migrate_data=None):
        post_method(context, instance_ref, dest, block_migration,
                            migrate_data)
        return

    def check_can_live_migrate_destination_cleanup(self, ctxt,
                                                   dest_check_data):
        return

    def check_can_live_migrate_destination(self, ctxt, instance_ref,
                                           src_compute_info, dst_compute_info,
                                           block_migration=False,
                                           disk_over_commit=False):
        return {}

    def check_can_live_migrate_source(self, ctxt, instance_ref,
                                      dest_check_data):
        return

    def finish_migration(self, context, migration, instance, disk_info,
                         network_info, image_meta, resize_instance,
                         block_device_info=None, power_on=True):
        return

    def confirm_migration(self, migration, instance, network_info):
        return

    def pre_live_migration(self, context, instance_ref, block_device_info,
                           network_info, disk, migrate_data=None):
        return

    def unfilter_instance(self, instance_ref, network_info):
        return

    def test_remove_vm(self, instance_name):
        """Removes the named VM, as if it crashed. For testing."""
        self.instances.pop(instance_name)

    def get_host_stats(self, refresh=False):
        """Return fake Host Status of ram, disk, network."""
        stats = []
        for nodename in _FAKE_NODES:
            host_status = self.host_status_base.copy()
            host_status['hypervisor_hostname'] = nodename
            host_status['host_hostname'] = nodename
            host_status['host_name_label'] = nodename
            stats.append(host_status)
        if len(stats) == 0:
            raise exception.NovaException("FakeDriver has no node")
        elif len(stats) == 1:
            return stats[0]
        else:
            return stats

    def host_power_action(self, host, action):
        """Reboots, shuts down or powers up the host."""
        return action

    def host_maintenance_mode(self, host, mode):
        """Start/Stop host maintenance window. On start, it triggers
        guest VMs evacuation.
        """
        if not mode:
            return 'off_maintenance'
        return 'on_maintenance'

    def set_host_enabled(self, host, enabled):
        """Sets the specified host's ability to accept new instances."""
        if enabled:
            return 'enabled'
        return 'disabled'

    def get_volume_connector(self, instance):
        return {'ip': '127.0.0.1', 'initiator': 'fake', 'host': 'fakehost'}

    def get_available_nodes(self, refresh=False):
        return _FAKE_NODES

    def instance_on_disk(self, instance):
        return False


class FakeVirtAPI(virtapi.VirtAPI):
    def provider_fw_rule_get_all(self, context):
        return db.provider_fw_rule_get_all(context)

    @contextlib.contextmanager
    def wait_for_instance_event(self, instance, event_names, deadline=300,
                                error_callback=None):
        # NOTE(danms): Don't actually wait for any events, just
        # fall through
        yield
